src/compiler/scala/tools/nsc/transform/Constructors.scala

```diff
@@ -659,66 +628,106 @@ abstract class Constructors extends Statics with Transform with ast.TreeDSL {
 
             initPhase += mkAssign(statSym, initializingRhs)
           }
+          !const // needs field unless it's a constant
+        }
 
-          defBuf += deriveValDef(stat)(_ => EmptyTree)
-
-        // all other statements go into the constructor
-        case _ => constrStatBuf += intoConstructor(impl.symbol, stat)
+        stat match {
+          // recurse on class definition, store in defBuf
+          case _: ClassDef => defBuf += new ConstructorTransformer(unit).transform(stat)
+
+          // Triage methods -- they all end up in the template --
+          // regular ones go to `defBuf`, secondary contructors go to `auxConstructorBuf`.
+          // The primary constructor is dealt with separately (we're massaging it here).
+          case _: DefDef if statSym.isPrimaryConstructor => ()
+          case _: DefDef if statSym.isConstructor => auxConstructorBuf += stat
+          case _: DefDef => defBuf += stat
+
+          // If a val needs a field, an empty valdef goes into the template.
+          // Except for lazy and ConstantTyped vals, the field is initialized by an assignment in:
+          //   - the class initializer (static),
+          //   - the constructor, before the super call (early initialized or a parameter accessor),
+          //   - the constructor, after the super call (regular val).
+          case ValDef(mods, _, _, rhs) =>
+            val memoized = moveEffectToCtor(mods, rhs)
+
+            if (memoized) defBuf += deriveValDef(stat)(_ => EmptyTree)
```

Comment:
For now, no change was intended, but it's entirely possible I made a mistake somewhere. The idea was that the old case from https://github.com/adriaanm/scala/commit/df850f0cb08742eb8af63893b028b6e367837530#diff-086eb87150c79730cf3cfe9cb8b12219L646 (` case _: ValDef if statSym.info.isInstanceOf[ConstantType] => ()`) is implemented here by returning `false` for from `moveEffectToCtor`, and thus no Valdef is emitted. You're right that the current compiler emits a field, but it doesn't seem to come from Constructors:

Both current scalac and the one from this PR do the following during Constructors for:

final.scala:
```
class C { final val C = 1 }
```

```
[[syntax trees at end of              constructors]] // final.scala
package <empty> {
  class C extends Object {
    final <stable> <accessor> def C(): Int(1) = 1;
    def <init>(): C = {
      C.super.<init>();
      ()
    }
  }
}
```

However, javap confirms your analysis:
```
Classfile /private/tmp/C.class
  Last modified Sep 1, 2015; size 552 bytes
  MD5 checksum a6c08e38b206f9e1dd72e29fc83ad9e3
  Compiled from "final.scala"
public class C
  SourceFile: "final.scala"
<snip>
{
  private final int C;
    descriptor: I
    flags: ACC_PRIVATE, ACC_FINAL
```